tests/test_density_matrix.py

```diff
@@ -0,0 +1,397 @@
+
+        backend = DensityMatrixBackend(pattern)
+        backend.add_nodes([0, 1, 2])
+        backend.entangle_nodes((0, 1))
+        backend.entangle_nodes((1, 2))
+        backend.measure(backend.pattern.seq[-4])
+
+        expected_matrix_1 = np.zeros((4, 4))
+        expected_matrix_1[0, 0] = 0.5
+        expected_matrix_1[0, 1] = 0.5
+        expected_matrix_1[1, 0] = 0.5
+        expected_matrix_1[1, 1] = 0.5
+        expected_matrix_2 = np.zeros((4, 4))
+        expected_matrix_2[2, 2] = 0.5
+        expected_matrix_2[2, 3] = -0.5
+        expected_matrix_2[3, 2] = -0.5
+        expected_matrix_2[3, 3] = 0.5
+
+        assert np.allclose(backend.state.rho, expected_matrix_1) or np.allclose(backend.state.rho, expected_matrix_2)
+
+    def test_correct_byproduct(self):
+        circ = Circuit(1)
+        circ.rx(0, np.pi / 2)
+        pattern = circ.transpile()
+
+        backend = DensityMatrixBackend(pattern)
+        backend.add_nodes([0, 1, 2])
+        backend.entangle_nodes((0, 1))
+        backend.entangle_nodes((1, 2))
+        backend.measure(backend.pattern.seq[-4])
+        backend.measure(backend.pattern.seq[-3])
+        backend.correct_byproduct(backend.pattern.seq[-2])
+        backend.correct_byproduct(backend.pattern.seq[-1])
+        backend.finalize()
+        rho = backend.state.rho
+
+        backend = StatevectorBackend(pattern)
+        backend.add_nodes([0, 1, 2])
+        backend.entangle_nodes((0, 1))
+        backend.entangle_nodes((1, 2))
+        backend.measure(backend.pattern.seq[-4])
+        backend.measure(backend.pattern.seq[-3])
+        backend.correct_byproduct(backend.pattern.seq[-2])
+        backend.correct_byproduct(backend.pattern.seq[-1])
+        backend.finalize()
+        psi = backend.state.psi
+
+        np.testing.assert_allclose(rho, np.outer(psi, psi.conj()))
+
+    def test_dephase(self):
+        def simulate(p, pattern):
+            backend = DensityMatrixBackend(pattern)
+            backend.add_nodes([0])
+            backend.dephase(p)
+            backend.add_nodes([1])
+            backend.dephase(p)
+            backend.add_nodes([2])
+            backend.dephase(p)
+            backend.entangle_nodes((0, 1))
+            backend.dephase(p)
+            backend.entangle_nodes((1, 2))
+            backend.dephase(p)
+            backend.measure(pattern.seq[-4])
+            backend.dephase(p)
+            backend.measure(pattern.seq[-3])
+            backend.dephase(p)
+            backend.correct_byproduct(pattern.seq[-2])
+            backend.dephase(p)
+            backend.correct_byproduct(pattern.seq[-1])
+            backend.dephase(p)
+            backend.finalize()
+            return backend
+
+        circ = Circuit(1)
+        circ.rx(0, np.pi / 4)
+        pattern = circ.transpile()
+        backend1 = simulate(0, pattern)
+        backend2 = simulate(1, pattern)
+        np.testing.assert_allclose(backend1.state.rho, backend2.state.rho)
```

Comment:
what is this testing?

Reply:
This is checking the concrete case where both resulting states becomes equal(no dephase case(p=0) and p=1 case).